lib/maths/analytics/CBoostedTreeHyperparameters.cc

```diff
@@ -402,52 +376,53 @@ void CBoostedTreeHyperparameters::initializeFineTuneSearch(double lossGap, std::
-                                                               std::size_t numberTrees) {
+void CBoostedTreeHyperparameters::checkIfCanSkipFineTuneSearch(double testLossVariance) {
+
     if (m_EarlyHyperparameterOptimizationStoppingEnabled && m_IncrementalTraining == false) {
+
+        auto toRestore = std::make_unique<common::CBayesianOptimisation>(
+            m_BayesianOptimization->boundingBox(), m_BayesianOptimization->restarts());
+        std::swap(toRestore, m_BayesianOptimization);
+
         // Add information about observed line search training runs to the GP.
         for (auto & [ parameters, loss ] : m_LineSearchHyperparameterLosses) {
-            TVector parameters_{
-                static_cast<TVector::TIndexType>(relevantParameters.size())};
-            for (std::size_t i = 0; i < relevantParameters.size(); ++i) {
-                parameters_(static_cast<TVector::TIndexType>(i)) =
-                    parameters(relevantParameters[i]);
+            TVector parameters_{static_cast<TVector::TIndexType>(
+                m_LineSearchRelevantParameters.size())};
+            for (std::size_t i = 0; i < m_LineSearchRelevantParameters.size(); ++i) {
+                parameters_(i) = parameters(m_LineSearchRelevantParameters[i]);
             }
-            m_BayesianOptimization->add(std::move(parameters_), loss, 0.0);
+            m_BayesianOptimization->add(std::move(parameters_), loss, testLossVariance);
         }
 
-        // Perform 3 additional rounds of kernel parameter optimization.
-        for (std::size_t i = 0; i < 3; ++i) {
-            m_BayesianOptimization->maximumLikelihoodKernel();
-        }
+        // Perform 3 rounds of kernel parameter optimization.
+        m_BayesianOptimization->maximumLikelihoodKernel(3);
 
         m_StopHyperparameterOptimizationEarly = this->optimisationMakingNoProgress();
+
         if (m_StopHyperparameterOptimizationEarly) {
             LOG_DEBUG(<< "Skipping fine-tune hyperparameters");
         } else {
             // Only reset Bayesian optimisation if we are going to fine tune or
             // else we won't be  able to compute hyperparameter importances.
-            this->resetBayesianOptimization();
+            std::swap(m_BayesianOptimization, toRestore);
         }
-        m_LineSearchHyperparameterLosses.clear();
-        m_LineSearchHyperparameterLosses.shrink_to_fit();
     }
-
-    // We purposely don't record the test loss because it isn't comparable with the
-    // value computed in fine tune. However, an estimate of the test train loss gap
-    // is needed if we incrementally train.
-    m_BestForestLossGap = lossGap;
-
-    // The stored number of trees is used for final train when we train on
-    // all the data and so can't measure when to stop.
-    std::size_t numberTreesToRestore{m_MaximumNumberTrees.value()};
-    m_MaximumNumberTrees.set(numberTrees);
-    this->saveCurrent();
-    m_MaximumNumberTrees.set(numberTreesToRestore);
```

Comment:
Why did this become irrelevant?

Reply:
I just moved it to `initializeFineTuneSearch`, so I could also call this function from `selectNext`.